lexer/lex.go

```diff
@@ -1,13 +1,318 @@
 	"github.com/NeowayLabs/abad/token"
 )
 
-type (
-	Tokval struct {
-		Type  token.Type
-		Value utf16.Str
+type Tokval struct {
+	Type  token.Type
+	Value utf16.Str
+	Line uint
+	Column uint
+}
+
+var EOF Tokval = Tokval{ Type: token.EOF }
+
+func (t Tokval) Equal(other Tokval) bool {
+	return t.Type == other.Type && t.Value.Equal(other.Value)
+}
+
+func (t Tokval) EqualPos(other Tokval) bool {
+	return t.Line == other.Line && t.Column == other.Column
+}
+
+// Lex will lex the given crappy JS code (utf16 yay) and provide a
+// stream of tokens as a result (the returned channel).
+//
+// The caller should iterate on the given channel until it is
+// closed indicating a EOF (or an error). Errors should be
+// handled by checking the type of the token.
+//
+// A goroutine will be started to lex the given code, if you
+// do not iterate the returned channel the goroutine will leak,
+// you MUST drain the provided channel.
+func Lex(code utf16.Str) <-chan Tokval {
+
+	tokens := make(chan Tokval)
+	
+	go func() {
+	
+		decodedCode := code.Runes()
+		currentState := newLexer(decodedCode).initialState
+		
+		for currentState != nil {
+			token, newState := currentState()
+			tokens <- token
+			currentState = newState
+		}
+		
+		close(tokens)
+	}()
+
+	return tokens
+}
+
+type lexer struct {
+	code []rune
+	position uint
+	line uint
+	column uint
+}
+
+type lexerState func() (Tokval, lexerState)
+
+func newLexer(code []rune) *lexer {
+	return &lexer {code:code, line: 1, column: 1}
+}
+
+func (l *lexer) initialState() (Tokval, lexerState) {
+
+	if l.isEOF() {
+		return EOF, nil
+	}
+		
+	if l.isInvalidRune() {
+		return l.illegalToken()
+	}
+	
+	if l.isPlusSign() {
+		// TODO: handle ++
+		return l.token(token.Plus), l.initialState
+	}
+	
+	if l.isMinusSign() {
+		// TODO: handle --
+		return l.token(token.Minus), l.initialState
+	}
+		
+	if l.isNumber() {
+		l.fwd()
+		return l.numberState()
+	}
+		
+	if l.isDot() {
+		l.fwd()
+		return l.realDecimalState()
+	}
+		
+	return l.identifierState()
+}
+
+func (l *lexer) numberState() (Tokval, lexerState) {
+
+	if l.isEOF() {
+		return l.token(token.Decimal), l.initialState
+	}
+	
+	if l.isDot() {
+		l.fwd()
+		return l.realDecimalState()
+	}
+	
+	if l.isNumber() {
+		l.fwd()
+		return l.decimalState(true)
+	}
+	
+	if l.isHexStart() {
+		l.fwd()
+		
+		if l.isEOF() {
+			return l.illegalToken()
+		}
+		
+		return l.hexadecimalState()
+	}
+	
+	if l.isExponentPartStart() {
+		l.fwd()
+		return l.exponentPartState()
+	}
+		
+	return l.illegalToken()
+}
+
+func (l *lexer) illegalToken() (Tokval, lexerState) {
+	return Tokval{
+		Type: token.Illegal,
+		Value: newStr(l.code),
+	}, nil
+}
+
+func (l *lexer) identifierState() (Tokval, lexerState) {
+
+	// TODO: obviously not complete
+	for !l.isEOF() {
+		l.fwd()
+	}
+		
+	return l.token(token.Ident), l.initialState
+}
+
+func (l *lexer) hexadecimalState() (Tokval, lexerState) {
+
+	for !l.isEOF() {
+		if !l.isHexadecimal() {
```

Comment:
not complete, right?

Reply:
Im not sure if it will ever be complete =P. On this particular case what it seems to be missing? (perhaps its something that Im not planning to fix =P)